### PR TITLE
FFI at CT

### LINKED ISSUE #9253: enable FFI at CT
## use cases
* there are so many, I don't think we need to justify the need. Just one example: enabling nre (or re) at CT (which isn't possible without FFI at CT because it uses PRCRE lib cf [irc](https://gitter.im/nim-lang/Nim?at=5bbbec33bbdc0b25050a011c) /cc @flaviut ) 

## discussion

**EDIT** 

in https://forum.nim-lang.org/t/4249#26456 @araq mentioned he had a working patch that allowed using FFI at CT for arbitrary importc procs, but it wasn't merged because it wasn't clear what to do with cases like these:

```c
void a(int* x; int L);
```
> Does the a write into x? If it does, we need to write back the changes performed to the blob x to the symbolic representation the VM uses.

However we can definitely think of solutions for this case, eg:
* pragma annotations to distinguish this
* disallow specific cases like this for now, but allow other cases that don't have this issue

## links
**EDIT**
* keeps popping up in IRC, eg: https://gitter.im/nim-lang/Nim?at=5bc3e7a3ef4afc4f2875952b
* https://forum.nim-lang.org/t/4249#26456 (Why can't the Nim VM use FFI?)
* https://github.com/nim-lang/Nim/issues/61#issuecomment-2619958 comment by @zah 
> Integration of libffi in the evaluation engine that will allow calling arbitrary C functions
* https://github.com/Araq/libffi libffi wrapper for Nim.
* comment from @araq https://forum.nim-lang.org/t/1743#10936
> The VM used to support the FFI (via libffi) and I can dig out this code again, but [...]

* http://www.chiark.greenend.org.uk/doc/libffi-dev/html/The-Basics.html

### PR DESCRIPTION
**Pull request by timotheecour**
just in time for the new year!
* fix #9253
* partially addresses #8405

[EDIT]
```nim
nimble install libffi
nim c -d:nimHasLibFFI compiler/nim.nim
compiler/nim c --experimental:compiletimeFFI -r tests/vm/tevalffi.nim
```

the test shows, as an example, that we can use these at CT:
```nim
proc c_exp(a: float64): float64 {.importc: "exp", header: "<math.h>".}
proc c_printf(frmt: cstring): cint {.importc: "printf", header: "<stdio.h>", varargs, discardable.}
proc c_snprintf*(buffer: pointer, buf_size: uint, format: cstring): cint {.importc: "snprintf", header: "<stdio.h>", varargs .}
proc c_malloc(size:uint):pointer {.importc:"malloc", header: "<stdlib.h>".}
proc c_free(p: pointer) {.importc:"free", header: "<stdlib.h>".}
```

sure, there may be edge cases not yet supported or buggy, but that's already super useful

## note
~~https://github.com/nim-lang/Nim/pull/10611 should be merged first (ignore it in the diff); it's needed because of https://github.com/nim-lang/Nim/issues/10632~~
EDIT: removed it

### CODE REVIEW

File: compiler/evalffi.nim [hunk 1]
```diff
@@ -10,41 +10,45 @@
...
 
 when defined(windows):
   const libcDll = "msvcrt.dll"
-else:
+elif defined(linux):
   const libcDll = "libc.so(.6|.5|)"
+elif defined(osx):
+  const libcDll = "/usr/lib/libSystem.dylib"
+else:
+  static: doAssert false
```
Comment (alaviss) [inline compiler/evalffi.nim]:
```suggestion
  {.error: "The FFI evaluator is not implemented on this platform".}
```

**Comment by dom96** [comment]
yay! One step closer to getting a fast Nim REPL :D

**Comment by krux02** [comment]
Please elaborate on the "edge cases not yet supported or buggy". That test is currently a bit minimal. Can you run an interactive SDL application with an OpenGL context in a window at compile time?

**Comment by zah** [comment]
I'm curious, does this fix https://github.com/nim-lang/Nim/issues/8405 as well?
If it does, I suggest enabling it by default because the issue above is quite a blocker for every day usage of the new hot code reloading branch.

**Comment by Araq** [comment]
The unwritten policy is that new features start behind an ``.experimental`` switch even if we all think "it's really solid and will cause no problems".

**Comment by timotheecour** [comment]
> The unwritten policy is that new features start behind an .experimental switch even if we all think "it's really solid and will cause no problems".

* `--experimental:allowFFI` does what you'd expect (enables/disables the feature)
* however it can't be localized via `push`: `{.push experimental: "allowFFI".}` does not work as I'm showing in tests/vm/tevalffi.nim (https://github.com/nim-lang/Nim/blob/dbea28d7bde654cd4b0705682e935abad8a2d3f9/tests/vm/tevalffi.nim#L9), since VM Context is different from regular Context

(still need to push my latest changes btw)

**Comment by alaviss** [comment]
> `--experimental:allowFFI` does what you'd expect (enables/disables the feature)

This might be intepreted that you would need this flag to use any kind of FFI in Nim. With that said, I can't come up with any better name.

File: compiler/options.nim [hunk 2]
```diff
@@ -126,6 +126,8 @@ type
     forLoopMacros,
     caseStmtMacros,
     codeReordering,
+    allowFFI, ## allow FFI at compile time; requires building nim with `-d:useFFI`
```
Comment (Araq) [inline compiler/options.nim]: ``compiletimeFFI``

**Comment by mratsim** [comment]
`AllowStaticFFI` or `AllowCTFFI`?

File: compiler/options.nim (hunk 2 above)
Comment (timotheecour) [reply compiler/options.nim]: done

**Comment by timotheecour** [comment]
ok finally getting back to this, rebased + added support for more code to work at CT, eg `printf`, `sprintf`, see tests.

@zah
> I'm curious, does this fix #8405 as well?

partially, we can add missing support in subsequent PRs.
eg: this doesn't work with `-d:useNimRtl`  but works with this PR
```nim
import strutils
# const a = "abc" in "abcdef" # dosen't yet work because `tyRange` isn't mapped in `mapType`
const a = "abc".countLines # this works fine
echo(a)
```

File: testament/categories.nim [hunk 3]
```diff
@@ -606,62 +614,70 @@ proc runJoinedTest(r: var TResults, cat: Category, testsDir: string) =
...
+      writeFile("outputExpected.txt", outputExpected)
+      discard execShellCmd("diff -uNdr outputExpected.txt outputGotten.txt")
+      echo "output different!"
+      # outputGotten.txt, outputExpected.txt not removed on purpose for debugging.
+      quit 1
+    else:
+      echo "output OK"
+      removeFile("outputGotten.txt")
+      removeFile(megatestFile)
+    #testSpec r, makeTest("megatest", options, cat)
```
Comment (krux02) [inline testament/categories.nim]: why is here so much diff in categories.nim? AFAIK this file doesn't need to be touched to test compile time FFI.

Comment (timotheecour) [reply testament/categories.nim]: @krux02 
I explained this in top-post: 
> #10611 should be merged first (ignore it in the diff); it's needed because of #10632

Comment (krux02) [reply testament/categories.nim]: No it's not needed. As far as I can tell #10611 is unrelated to this PR. Please disconnect this PR from the other one.

Comment (timotheecour) [reply testament/categories.nim]: > No it's not needed.

I just removed it.
As a result, as explained in #10611, I can't test this PR using my own travis instance, so my previous workflow doesn't work anymore:
* A1: test a commit locally; if green:
* A2: push on a private branch; if localy travis is green:
* A3: push on PR branch; this runs on shared travis resources

the point of A2 being to avoid wasting shared travis resources. So I have to keep rebasing/removing that PR now

Comment (krux02) [reply testament/categories.nim]: I noted in a previous comment to elaborate what local travis is. This pretty much explais it to me. Still it won't harm if you describe it in the PR anyway. But no you don't need to "keep rebasing/removing", you just push to the branch of the PR and work on shared travis resources. That will simplify your life.

**Comment by Araq** [comment]
Merging so that we don't lose this work but I'll disable it afterwards.
